### ISSUE TITLE
Implement optimality tightening

### ISSUE THREAD

**Issue by Kaixhin** [Kaixhin/Atari#60]
[Learning to Play in a Day: Faster Deep Reinforcement Learning by Optimality Tightening](https://arxiv.org/abs/1611.01606) potentially speeds up Q-learning by an order of magnitude! Apparently not too hard to implement either.

**Comment by petrosgk** [comment]
I gave it a shot, however I am not sure how the discounted reward R is supposed to be used and I also need to check if future and past k-transitions are valid

https://github.com/petrosgk/Atari/tree/opt-tightening

**Comment by Kaixhin** [comment]
Awesome - I'll try and have a look soon or next week! Would you be able to test it to try and replicate one of the results from the paper?

I started on this myself as well, so will see how our implementations compare.

**Comment by Aeroone** [comment]
Hi, have you reproduced that optimality tightening results? I have tried some games based on tensorflow and openai gym but the results seem much worse than the papers' results. I am not sure whether I misunderstand something or miss some tricks in the paper. It seems that the paper doesn't include everything about their works.

**Comment by DanielTea** [comment]
Does anyone know wether they have published the source code for optimal tightening, from the paper?

**Comment by Aeroone** [comment]
No, they haven't published their code as far as I know. The tricks they use are not hard to implement but I can not still achieve their performance.

**Comment by Aeroone** [comment]
No, they haven't published their source code as far as I know. The tricks they use are not hard to implement but I can not still achieve their performance.

**Comment by petrosgk** [comment]
I have tried implementing optimality tightening (see earlier post) but the results I get are also much worse than the paper's.

**Comment by Kaixhin** [comment]
In my experience the smallest details in a paper can be key to reproducing results - and these may be missing or ambiguous. If anyone is reasonably confident in their implementation, you should try contacting one of the authors with specific questions.

**Comment by tadashiK** [comment]
@petrosgk Do you have a plot of a learning curve? I read the paper, and I am a bit confused with their results since DQN also learns very quickly in their paper. In Figure 3, DQN attains 2000 points around 10M frames in Zaxxon. However, in prioritized experience paper, DQN's score is almost 0 around 10M frames. I am suspecting that the authors misunderstand the meaning of "frames" in DQN as frames the agent observes. Then, it explains very quick learning speed of DQN in their paper. If your algorithm learns around 5 times slower than their result, your algorithm might be working correctly.

**Comment by ShibiHe** [comment]
Hi guys,
I have released the code at https://github.com/ShibiHe/Q-Optimality-Tightening. Please have a look.

Best,
Shibi
